### ISSUE TITLE
[Issue]: RX 9060 XT gfx1200 gfxhub page fault during LoRA training, mitigated by disabling PyTorch HIP memory caching

### ISSUE THREAD

**Issue by acranerafael** [ROCm/ROCm#6647]
### Problem Description

I'm experiencing reproducible [gfxhub] page faults during LoRA fine-tuning with PyTorch + Unsloth on an AMD Radeon RX 9060 XT (RDNA4 / gfx1200) using ROCm 7.14.

After controlled testing, I obtained a deterministic reproduction:

- With gradient checkpointing disabled and a fixed data order, 7/7 training runs failed at exactly step 105/344.
- With only PYTORCH_NO_HIP_MEMORY_CACHING=1 added, 3/3 identical training runs completed all 344/344 steps successfully.
- No new gfxhub events were recorded in dmesg during any of those three successful runs.
- I verified that the PyTorch caching allocator was actually disabled: after allocating a GPU tensor, torch.cuda.memory_stats() reported reserved=0 and segments=0.

Typical kernel fault:

[gfxhub] page fault
Faulty UTCL2 client ID: TCP (0x8)
PERMISSION_FAULTS: 0x3
MAPPING_ERROR: 0x1
WALKER_ERROR: 0x5

The fault occurs during LoRA backward GEMMs. Observed shapes include:

transA = N
transB = T
M = 1024 / 2048 / 3072
N = 16 (LoRA rank)
K ~= 243-343

The observed kernel family contains:

Cijk_Ailk_Bjlk_...
MT64x64x32_MI16x16x1
ISA1200

I created an isolated reproducer for this GEMM family. It executed 20.86 million GEMMs across 18 M x K combinations with N=16 for 420 seconds without a single fault or new gfxhub event. This is roughly 300x more invocations than occur during one complete training run.

Other controlled experiments:

- hipBLASLt: fault occurs
- rocBLAS (TORCH_BLAS_PREFER_HIPBLASLT=0): fault still occurs
- BF16: fault occurs
- FP16: 4/5 completed, 1/5 produced the same gfxhub fault
- Unsloth gradient checkpointing enabled: intermittent faults at different steps
- Gradient checkpointing disabled: deterministic 7/7 fault at step 105
- Changing only Hugging Face data_seed moved the fault:
  - data_seed=3407 -> step 105
  - data_seed=1234 -> step 211
  - data_seed=4321 -> step 34

Allocator instrumentation immediately before the deterministic fault showed no OOM or obvious memory pressure:

allocated ~= 1279 MB
reserved ~= 2932 MB
segments = 89
inactive_split ~= 56-61 MB
allocation retries = 0
OOMs = 0

Reserved memory had stopped growing approximately 41 steps before the fault.

Workaround results with PYTORCH_NO_HIP_MEMORY_CACHING=1:

Run 1: 344/344 OK - 443 s - no gfxhub event
Run 2: 344/344 OK - 430 s - no gfxhub event
Run 3: 344/344 OK - 439 s - no gfxhub event

Baseline without the workaround:

7/7 failures at exactly step 105/344.

The workaround has a significant performance cost: approximately 437 seconds versus 124 seconds (~3.5x slower).

I am not claiming that the PyTorch caching allocator itself is necessarily the root cause. Disabling memory caching also changes execution timing significantly, so this could alternatively be a synchronization/race condition elsewhere in the ROCm/runtime/driver stack.

This appears related to ROCm issue #6600, which reports deterministic gfxhub page faults during LoRA/QLoRA training on RDNA4/gfx1201 where the fault step also changes with data order.

The important additional observation here is that this reproduces on RX 9060 XT / gfx1200 and that PYTORCH_NO_HIP_MEMORY_CACHING=1 provides a repeatable mitigation.

I can provide the complete dmesg logs, isolated GEMM reproducer, deterministic training reproducer, allocator instrumentation logs, Docker configuration, and full package versions if needed.

### Operating System

Ubuntu 24.04 LTS

### CPU

AMD Ryzen 7 5700X 8-Core Processor

### GPU

AMD Radeon RX 9060 XT 16 GB (RDNA4 / gfx1200)

### ROCm Version

ROCm 7.14

### Steps to Reproduce

1. Run PyTorch + Unsloth with ROCm 7.14 on an AMD Radeon RX 9060 XT (gfx1200) inside an Ubuntu 24.04 Docker environment.

2. Fine-tune Qwen3-0.6B using LoRA with:
   - LoRA rank: 16
   - BF16
   - batch size: 2
   - gradient accumulation: 2
   - gradient checkpointing: disabled
   - fixed model seed: 3407
   - fixed data_seed: 3407
   - 342 training examples

3. Run the training normally with the PyTorch HIP caching allocator enabled.

Expected behavior:
Training should complete normally.

Actual behavior:
The training reproducibly triggers a gfxhub page fault.

Using the same configuration and data order:
7/7 runs failed at exactly step 105/344.

dmesg reports:

[gfxhub] page fault
Faulty UTCL2 client ID: TCP (0x8)
PERMISSION_FAULTS: 0x3
MAPPING_ERROR: 0x1
WALKER_ERROR: 0x5

4. Run the exact same training with:

PYTORCH_NO_HIP_MEMORY_CACHING=1

Result:
3/3 runs completed all 344/344 steps successfully with no new gfxhub events in dmesg.

The workaround was verified to actually disable the PyTorch caching allocator: after a real GPU tensor allocation, torch.cuda.memory_stats() reported reserved=0 and segments=0.

Note: disabling memory caching makes this workload approximately 3.5x slower (~437 s vs ~124 s).

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```text
ROCk module is loaded
=====================
HSA System Attributes
=====================
Runtime Version:         1.21
Runtime Ext Version:     1.26
System Timestamp Freq.:  1000.000000MHz
Sig. Max Wait Duration:  18446744073709551615 (0xFFFFFFFFFFFFFFFF) (timestamp count)
Machine Model:           LARGE
System Endianness:       LITTLE
Mwaitx:                  ENABLED
XNACK enabled:           NO
DMAbuf Support:          YES
VMM Support:             YES
Fabric Support:          NO

==========
HSA Agents
==========
*******
Agent 1
*******
  Name:                    AMD Ryzen 7 5700X 8-Core Processor
  Uuid:                    [REDACTED]
  Marketing Name:          AMD Ryzen 7 5700X 8-Core Processor
  Vendor Name:             CPU
  Feature:                 None specified
  Profile:                 FULL_PROFILE
  Float Round Mode:        NEAR
  Max Queue Number:        0(0x0)
  Queue Min Size:          0(0x0)
  Queue Max Size:          0(0x0)
  Queue Type:              MULTI
  Node:                    0
  Device Type:             CPU
  Cache Info:
    L1:                      32768(0x8000) KB
  Chip ID:                 0(0x0)
  ASIC Revision:           0(0x0)
  Cacheline Size:          64(0x40)
  Max Clock Freq. (MHz):   4665
  BDFID:                   0
  Internal Node ID:        0
  Compute Unit:            16
  SIMDs per CU:            0
  Shader Engines:          0
  Shader Arrs. per Eng.:   0
  WatchPts on Addr. Ranges:1
  Memory Properties:
  Features:                None
  Pool Info:
    Pool 1
      Segment:                 GLOBAL; FLAGS: FINE GRAINED
      Size:                    31732620(0x1e4338c) KB
      Allocatable:             TRUE
      Alloc Granule:           4KB
      Alloc Recommended Granule:4KB
      Alloc Alignment:         4KB
      Accessible by all:       TRUE
    Pool 2
      Segment:                 GLOBAL; FLAGS: EXTENDED FINE GRAINED
      Size:                    31732620(0x1e4338c) KB
      Allocatable:             TRUE
      Alloc Granule:           4KB
      Alloc Recommended Granule:4KB
      Alloc Alignment:         4KB
      Accessible by all:       TRUE
    Pool 3
      Segment:                 GLOBAL; FLAGS: KERNARG, FINE GRAINED
      Size:                    31732620(0x1e4338c) KB
      Allocatable:             TRUE
      Alloc Granule:           4KB
      Alloc Recommended Granule:4KB
      Alloc Alignment:         4KB
      Accessible by all:       TRUE
    Pool 4
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED
      Size:                    31732620(0x1e4338c) KB
      Allocatable:             TRUE
      Alloc Granule:           4KB
      Alloc Recommended Granule:4KB
      Alloc Alignment:         4KB
      Accessible by all:       TRUE
  ISA Info:

*******
Agent 2
*******
  Name:                    gfx1200
  Uuid:                    [REDACTED]
  Marketing Name:          AMD Radeon RX 9060 XT
  Vendor Name:             AMD
  Feature:                 KERNEL_DISPATCH
  Profile:                 BASE_PROFILE
  Float Round Mode:        NEAR
  Max Queue Number:        128(0x80)
  Queue Min Size:          64(0x40)
  Queue Max Size:          131072(0x20000)
  Queue Type:              MULTI
  Node:                    1
  Device Type:             GPU
  Cache Info:
    L1:                      32(0x20) KB
    L2:                      4096(0x1000) KB
    L3:                      32768(0x8000) KB
  Chip ID:                 30096(0x7590)
  ASIC Revision:           1(0x1)
  Cacheline Size:          256(0x100)
  Max Clock Freq. (MHz):   2780
  BDFID:                   1792
  Internal Node ID:        1
  Compute Unit:            32
  SIMDs per CU:            2
  Shader Engines:          2
  Shader Arrs. per Eng.:   2
  WatchPts on Addr. Ranges:4
  Coherent Host Access:    FALSE
  Memory Properties:
  Features:                KERNEL_DISPATCH
  Fast F16 Operation:      TRUE
  Wavefront Size:          32(0x20)
  Workgroup Max Size:      1024(0x400)
  Workgroup Max Size per Dimension:
    x                        1024(0x400)
    y                        1024(0x400)
    z                        1024(0x400)
  Max Waves Per CU:        32(0x20)
  Max Work-item Per CU:    1024(0x400)
  Grid Max Size:           4294967295(0xffffffff)
  Grid Max Size per Dimension:
    x                        4294967295(0xffffffff)
    y                        65535(0xffff)
    z                        65535(0xffff)
  Max fbarriers/Workgrp:   32
  Packet Processor uCode:: 178
  SDMA engine uCode::      662
  IOMMU Support::          None
  Pool Info:
    Pool 1
      Segment:                 GLOBAL; FLAGS: COARSE GRAINED
      Size:                    16695296(0xfec000) KB
      Allocatable:             TRUE
      Alloc Granule:           4KB
      Alloc Recommended Granule:2048KB
      Alloc Alignment:         4KB
      Accessible by all:       FALSE
    Pool 2
      Segment:                 GROUP
      Size:                    64(0x40) KB
      Allocatable:             FALSE
      Alloc Granule:           0KB
      Alloc Recommended Granule:0KB
      Alloc Alignment:         0KB
      Accessible by all:       FALSE
  ISA Info:
    ISA 1
      Name:                    amdgcn-amd-amdhsa--gfx1200
      Machine Models:          HSA_MACHINE_MODEL_LARGE
      Profiles:                HSA_PROFILE_BASE
      Default Rounding Mode:   NEAR
      Fast f16:                TRUE
      Workgroup Max Size:      1024(0x400)
      Workgroup Max Size per Dimension:
        x                        1024(0x400)
        y                        1024(0x400)
        z                        1024(0x400)
      Grid Max Size:           4294967295(0xffffffff)
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)
        y                        65535(0xffff)
        z                        65535(0xffff)
      FBarrier Max Size:       32
    ISA 2
      Name:                    amdgcn-amd-amdhsa--gfx12-generic
      Machine Models:          HSA_MACHINE_MODEL_LARGE
      Profiles:                HSA_PROFILE_BASE
      Default Rounding Mode:   NEAR
      Fast f16:                TRUE
      Workgroup Max Size:      1024(0x400)
      Workgroup Max Size per Dimension:
        x                        1024(0x400)
        y                        1024(0x400)
        z                        1024(0x400)
      Grid Max Size:           4294967295(0xffffffff)
      Grid Max Size per Dimension:
        x                        2147483647(0x7fffffff)
        y                        65535(0xffff)
        z                        65535(0xffff)
      FBarrier Max Size:       32

*** Done ***
```

GPU UUID redacted for privacy.
</details>

### Environment details

The issue occurs inside a Docker container on Ubuntu 24.04 with ROCm 7.14 and an AMD Radeon RX 9060 XT (`gfx1200`).

Training stack:

- PyTorch: 2.10.0
- Triton: 3.7.1
- Unsloth: 2026.8.21
- TRL: 0.24.0
- Transformers: 5.5.0
- GPU architecture: RDNA4 / gfx1200
- Training framework: PyTorch + Unsloth
- Containerized environment: Docker

### Relevant environment variable / workaround

The currently confirmed mitigation is:

```bash
PYTORCH_NO_HIP_MEMORY_CACHING=1
```

The variable is applied inside the Docker container.

Its effect was independently verified with `torch.cuda.memory_stats()` after a real GPU allocation:

```text
reserved = 0
segments = 0
```

With this workaround:

```text
Run 1: 344/344 OK — 443 s — no gfxhub event
Run 2: 344/344 OK — 430 s — no gfxhub event
Run 3: 344/344 OK — 439 s — no gfxhub event
```

Without it, the deterministic reference configuration failed:

```text
7/7 runs failed at step 105/344
```

### dmesg fault signature

The failures produce an AMDGPU gfxhub page fault similar to:

```text
[gfxhub] page fault
Faulty UTCL2 client ID: TCP (0x8)
PERMISSION_FAULTS: 0x3
MAPPING_ERROR: 0x1
WALKER_ERROR: 0x5
```

No corresponding gfxhub event was recorded during any of the three successful runs with `PYTORCH_NO_HIP_MEMORY_CACHING=1`.

### Performance impact

Disabling the PyTorch HIP caching allocator has a substantial performance cost:

```text
Normal allocator: ~124 s
Caching disabled: ~437 s average
```

Approximately 3.5x slower for this workload.

Therefore, this is currently considered a mitigation rather than a satisfactory long-term solution.

### Related reports

This may be related to other RDNA4 / gfx12xx gfxhub page-fault reports, particularly:

- ROCm #6600 — similar LoRA/QLoRA behavior on gfx1201 where changing data order changes the fault step.
- ROCm #6413 — surfaced by GitHub as a potential related/duplicate gfx1201 report.

I am reporting this separately because this case reproduces on **gfx1200 / Radeon RX 9060 XT** and includes a repeatable mitigation using `PYTORCH_NO_HIP_MEMORY_CACHING=1`.

### Additional diagnostic data available

I can provide if requested:

- complete `dmesg` logs from failing runs;
- full faulting kernel names;
- `HIPBLASLT_LOG_LEVEL=2` logs;
- deterministic training reproduction script;
- isolated GEMM reproduction script;
- PyTorch allocator instrumentation logs;
- Docker Compose configuration;
- training configuration and exact reproduction parameters.

**Comment by harkgill-amd** [comment]
> Related reports
This may be related to other RDNA4 / gfx12xx gfxhub page-fault reports, particularly:
ROCm https://github.com/ROCm/legacy-rocm-build/issues/6600 — similar LoRA/QLoRA behavior on gfx1201 where changing data order changes the fault step.
ROCm https://github.com/ROCm/legacy-rocm-build/issues/6413 — surfaced by GitHub as a potential related/duplicate gfx1201 report.

Ya, this looks identical to both of those issues. #6600 has already been confirmed to be resolved with https://github.com/ROCm/rocm-libraries/pull/8909 which is already in the latest nightlies. Can you give the nightlies a try on your end https://github.com/ROCm/TheRock/blob/main/RELEASES.md#installing-multi-arch-python-packages.

Based on the workload and fail signature here, I'm pretty confident the aforementioned fix also resolves your failures here.

**Comment by harkgill-amd** [comment]
Closing this out for now. Please give the suggestions in the above comment a try when you get a chance.
